File: examples/webhook-listener.mjs
/*
 * Copyright (c) 2026, WSO2 LLC. (https://www.wso2.com).
 *
 * WSO2 LLC. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { createHmac, timingSafeEqual } from 'node:crypto';
import { createServer } from 'node:http';
import { pathToFileURL } from 'node:url';

function log(level, message) {
  const ts = new Date().toISOString().replace('T', ' ').substring(0, 19);
  console.log(`[${ts}] [${level}] ${message}`);
}

/**
 * Creates the HTTP request handler for subscription verification and event delivery.
 * @param {object} config - Configuration object { secret?, path? }
 */
export function createReceiver(config = {}) {
  const sharedSecret = config.secret ?? process.env.SHARED_SECRET ?? '';
  const callbackPath = config.path ?? process.env.CALLBACK_PATH ?? '/dpdp/events';

  return createServer(async (req, res) => {
    const reply = (status, body = '', contentType = 'text/plain') => {
      res.writeHead(status, { 'Content-Type': contentType, 'Cache-Control': 'no-store' });
      res.end(body);
    };

    const url = new URL(req.url, 'http://receiver.invalid');
    log('INFO', `--> Incoming ${req.method} ${url.pathname} from ${req.socket.remoteAddress || 'client'}`);

    if (url.pathname !== callbackPath) {
      log('WARN', `<-- Route not found: '${url.pathname}' (expected '${callbackPath}'). Returning HTTP 404.`);
      return reply(404, 'Not Found');
    }
    if (req.method !== 'POST') {
      log('WARN', `<-- Method '${req.method}' not allowed. Returning HTTP 405.`);
      return reply(405, 'Method Not Allowed');
    }
    if (req.headers['content-type']?.split(';')[0].trim() !== 'application/json') {
      log('WARN', `<-- Invalid Content-Type '${req.headers['content-type']}' (expected application/json). Returning HTTP 415.`);
      return reply(415, 'Unsupported Media Type');
    }

    try {
      const chunks = [];
      let size = 0;
      for await (const chunk of req) {
        size += chunk.length;
        if (size > 1024 * 1024) {
          log('WARN', '<-- Request body exceeds 1 MiB limit. Returning HTTP 413.');
          return reply(413, 'Payload Too Large');
        }
        chunks.push(chunk);
      }
      const raw = Buffer.concat(chunks);
      const message = JSON.parse(raw.toString('utf8'));

      // 1. Subscription Verification Handshake
      if (message.type === 'subscription.verification') {
        log('INFO', '================ [SUBSCRIPTION VERIFICATION] ================');
        log('INFO', `Verification Payload:\n${JSON.stringify(message, null, 2)}`);

        if (!message.challenge || typeof message.challenge !== 'string') {
          log('WARN', '<-- Verification rejected: missing challenge. Returning HTTP 400.');
          return reply(400, 'Invalid challenge');
        }

        log('INFO', `Verified subscription '${message.subscriptionId}' for topics: ${JSON.stringify(message.topics)}`);
        log('INFO', `<-- Returned HTTP 200 OK with challenge: '${message.challenge}'`);
        log('INFO', '=============================================================');
        return reply(200, message.challenge, 'text/plain; charset=utf-8');
      }

      // 2. Event Delivery
      const deliveryId = req.headers['delivery-id'] ?? 'unknown';
      const eventSignature = req.headers['event-signature'] ?? '';

      log('INFO', '==================== [EVENT DELIVERY] ====================');
      log('INFO', `Delivery-Id:     ${deliveryId}`);
      log('INFO', `Event-Signature: ${eventSignature || '(none)'}`);

      // Verify HMAC-SHA256 signature if shared secret is provided
      if (sharedSecret) {
        if (!eventSignature.startsWith('sha256=')) {
          log('ERROR', '<-- Missing or invalid Event-Signature header format. Returning HTTP 401.');
          return reply(401, 'Invalid Event-Signature header');
        }
        const expectedHmac = `sha256=${createHmac('sha256', sharedSecret).update(raw).digest('hex')}`;
        const bufExpected = Buffer.from(expectedHmac, 'utf8');
        const bufReceived = Buffer.from(eventSignature, 'utf8');

        if (bufExpected.length !== bufReceived.length || !timingSafeEqual(bufExpected, bufReceived)) {
          log('ERROR', '<-- HMAC-SHA256 signature verification failed. Returning HTTP 401.');
          return reply(401, 'HMAC-SHA256 signature mismatch');
        }
        log('INFO', 'HMAC-SHA256 signature verified successfully.');
      } else {
        log('INFO', 'SHARED_SECRET not set; skipped HMAC signature check.');
      }

      // Decode compact JWS claims if signedPayload is present
      if (message.signedPayload && typeof message.signedPayload === 'string') {
        log('INFO', `Raw Signed JWS:\n${message.signedPayload}`);
        try {
          const parts = message.signedPayload.split('.');
          if (parts.length >= 2) {
            const claims = JSON.parse(Buffer.from(parts[1], 'base64url').toString('utf8'));
            log('INFO', `Decoded Event Payload:\n${JSON.stringify(claims, null, 2)}`);
          }
        } catch (e) {
          log('WARN', `Could not decode JWS claims: ${e.message}`);
        }
      } else {
        log('INFO', `Event Body:\n${JSON.stringify(message, null, 2)}`);
      }

      log('INFO', '<-- Returned HTTP 202 Accepted');
      log('INFO', '==========================================================');
      return reply(202, JSON.stringify({ status: 'accepted' }), 'application/json');

    } catch (err) {
      log('ERROR', `<-- Malformed JSON body or request error: ${err.message}. Returning HTTP 400.`);
      return reply(400, 'Bad Request');
    }
  });
}

function main() {
  const port = Number(process.env.PORT || 8443);
  const host = process.env.HOST || '127.0.0.1';
  const callbackPath = process.env.CALLBACK_PATH || '/dpdp/events';
  const sharedSecret = process.env.SHARED_SECRET || '';

  const server = createReceiver({ secret: sharedSecret, path: callbackPath });

  server.listen(port, host, () => {
    console.log(`=============================================================`);
    console.log(`  DPDP Reference Webhook Listener (Node.js)`);
    console.log(`  Listening on: http://${host}:${port}${callbackPath}`);
    if (sharedSecret) {
      console.log(`  HMAC Verification: ENABLED (shared secret configured)`);
    } else {
      console.log(`  HMAC Verification: DISABLED (set SHARED_SECRET to enable)`);
    }
    console.log(`  Press Ctrl+C to stop.`);
    console.log(`=============================================================`);
  });

  for (const signal of ['SIGINT', 'SIGTERM']) {
    process.on(signal, () => {
      console.log('\nShutting down listener.');
      server.close();
    });
  }
}

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  main();
}

File: examples/webhook-listener.py
#!/usr/bin/env python3
"""
DPDP Reference Webhook Listener (Python Standard Library - Zero External Dependencies)

A simple, lightweight receiver for verifying subscriptions and inspecting delivered events.
Logs all verification challenges and event payloads directly to the console.

Requirements:
  Python 3.9+ (no pip packages or virtual environment required)

Usage:
  # Optional: set shared secret to verify HMAC signatures
  export SHARED_SECRET="my-shared-secret"

  # Optional overrides (defaults: PORT=8443, HOST=127.0.0.1, CALLBACK_PATH=/dpdp/events)
  export PORT=8443
  export CALLBACK_PATH="/dpdp/events"

  python3 docs/static/examples/webhook-listener.py
"""

import sys
import os
import json
import base64
import hmac
import hashlib
from http.server import HTTPServer, BaseHTTPRequestHandler
from datetime import datetime

PORT = int(os.environ.get("PORT", "8443"))
HOST = os.environ.get("HOST", "127.0.0.1")
CALLBACK_PATH = os.environ.get("CALLBACK_PATH", "/dpdp/events")
SHARED_SECRET = os.environ.get("SHARED_SECRET", "").strip()


def log(level, message):
    ts = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    print(f"[{ts}] [{level}] {message}", flush=True)


class WebhookRequestHandler(BaseHTTPRequestHandler):
    server_version = "DPDPWebhookListener/1.0"

    def _handle_unsupported_method(self):
        log("INFO", f"--> Incoming {self.command} {self.path} from {self.client_address[0]}")
        log("WARN", f"<-- Method '{self.command}' not allowed. Webhook endpoints accept POST only. Returning HTTP 405.")
        self.send_error(405, "Method Not Allowed")

    def __getattr__(self, name):
        # Catch-all for any non-POST HTTP method (GET, PUT, DELETE, TRACE, CONNECT, etc.)
        if name.startswith("do_"):
            return self._handle_unsupported_method
        raise AttributeError(name)

    def do_POST(self):
        log("INFO", f"--> Incoming POST {self.path} from {self.client_address[0]}")

        # 1. Validate route path
        if self.path.split("?")[0] != CALLBACK_PATH:
            log("WARN", f"<-- Route not found: '{self.path}' (expected '{CALLBACK_PATH}'). Returning HTTP 404.")
            self.send_error(404, "Unknown path")
            return

        # 2. Check Content-Type
        content_type = self.headers.get("Content-Type", "")
        if "application/json" not in content_type:
            log("WARN", f"<-- Unsupported Content-Type: '{content_type}' (expected 'application/json'). Returning HTTP 415.")
            self.send_error(415, "Content-Type must be application/json")
            return

        # 3. Read raw request body
        try:
            content_length = int(self.headers.get("Content-Length", 0))
        except (ValueError, TypeError):
            log("WARN", "<-- Invalid Content-Length header. Returning HTTP 400.")
            self.send_error(400, "Invalid Content-Length")
            return

        if content_length > 1024 * 1024:
            log("WARN", "<-- Request body exceeds 1 MiB limit. Returning HTTP 413.")
            self.send_error(413, "Payload too large (max 1 MiB)")
            return

        raw_body = self.rfile.read(content_length)

        try:
            message = json.loads(raw_body.decode("utf-8"))
        except Exception:
            log("ERROR", "<-- Malformed JSON body. Returning HTTP 400.")
            self.send_error(400, "Malformed JSON body")
            return

        # 4. Handle Subscription Verification
        if isinstance(message, dict) and message.get("type") == "subscription.verification":
            self.handle_verification(message)
            return

        # 5. Handle Event Delivery
        self.handle_event_delivery(raw_body, message)

    def handle_verification(self, message):
        log("INFO", "================ [SUBSCRIPTION VERIFICATION] ================")
        log("INFO", f"Verification Payload:\n{json.dumps(message, indent=2)}")

        challenge = message.get("challenge")
        if not challenge or not isinstance(challenge, str):
            log("WARN", "<-- Verification rejected: invalid or missing challenge. Returning HTTP 400.")
            self.send_error(400, "Invalid or missing challenge")
            return

        sub_id = message.get("subscriptionId", "unknown")
        topics = message.get("topics", [])

        # Respond HTTP 200 with raw challenge text
        body_bytes = challenge.encode("utf-8")
        self.send_response(200)
        self.send_header("Content-Type", "text/plain; charset=utf-8")
        self.send_header("Content-Length", str(len(body_bytes)))
        self.send_header("Cache-Control", "no-store")
        self.end_headers()
        self.wfile.write(body_bytes)

        log("INFO", f"Verified subscription '{sub_id}' for topics: {topics}")
        log("INFO", f"<-- Returned HTTP 200 OK with challenge: '{challenge}'")
        log("INFO", "=============================================================")

    def handle_event_delivery(self, raw_body, message):
        delivery_id = self.headers.get("Delivery-Id", "unknown")
        event_signature = self.headers.get("Event-Signature", "")

        log("INFO", "==================== [EVENT DELIVERY] ====================")
        log("INFO", f"Delivery-Id:     {delivery_id}")
        log("INFO", f"Event-Signature: {event_signature or '(none)'}")

        # Check HMAC-SHA256 signature if SHARED_SECRET is configured
        if SHARED_SECRET:
            if not event_signature.startswith("sha256="):
                log("ERROR", "<-- Missing or invalid Event-Signature header format. Returning HTTP 401.")
                self.send_error(401, "Missing or invalid Event-Signature header format")
                return

            expected_hmac = hmac.new(SHARED_SECRET.encode("utf-8"), raw_body, hashlib.sha256).hexdigest()
            received_hmac = event_signature[len("sha256="):].strip()

            if not hmac.compare_digest(expected_hmac.lower(), received_hmac.lower()):
                log("ERROR", "<-- HMAC-SHA256 signature verification failed. Returning HTTP 401.")
                self.send_error(401, "HMAC-SHA256 signature mismatch")
                return

            log("INFO", "HMAC-SHA256 signature verified successfully.")
        else:
            log("INFO", "SHARED_SECRET not set; skipped HMAC signature check.")

        # If signedPayload is present, decode compact JWS claims for inspection
        if isinstance(message, dict) and "signedPayload" in message:
            log("INFO", f"Raw Signed JWS:\n{message['signedPayload']}")
            try:
                parts = message["signedPayload"].split(".")
                if len(parts) >= 2:
                    padded = parts[1] + "=" * ((4 - len(parts[1]) % 4) % 4)
                    claims = json.loads(base64.urlsafe_b64decode(padded).decode("utf-8"))
                    log("INFO", f"Decoded Event Payload:\n{json.dumps(claims, indent=2)}")
            except Exception as e:
                log("WARN", f"Could not decode JWS claims: {e}")
        else:
            log("INFO", f"Event Body:\n{json.dumps(message, indent=2)}")

        # Respond HTTP 202 Accepted
        self.send_response(202)
        self.send_header("Content-Type", "application/json")
        self.send_header("Cache-Control", "no-store")
        self.end_headers()
        self.wfile.write(b'{"status":"accepted"}')
        log("INFO", "<-- Returned HTTP 202 Accepted")
        log("INFO", "==========================================================")

    def log_message(self, format, *args):
        # Capture unhandled internal server messages (e.g. syntax errors or unhandled HTTP verbs)
        msg = format % args
        if any(msg.startswith(prefix) for prefix in ("code 4", "code 5", "Unsupported")):
            log("WARN", f"[HTTP Server] {msg}")


def main():
    server_address = (HOST, PORT)
    httpd = HTTPServer(server_address, WebhookRequestHandler)
    print(f"=============================================================")
    print(f"  DPDP Reference Webhook Listener")
    print(f"  Listening on: http://{HOST}:{PORT}{CALLBACK_PATH}")
    if SHARED_SECRET:
        print(f"  HMAC Verification: ENABLED (shared secret configured)")
    else:
        print(f"  HMAC Verification: DISABLED (set SHARED_SECRET to enable)")
    print(f"  Press Ctrl+C to stop.")
    print(f"=============================================================")
    try:
        httpd.serve_forever()
    except KeyboardInterrupt:
        print("\nShutting down listener.")
        httpd.server_close()


if __name__ == "__main__":
    main()
